Lib/test/test_traceback.py

```diff
@@ -2800,6 +3221,87 @@ def test_all(self):
                 expected.add(name)
         self.assertCountEqual(traceback.__all__, expected)
 
+    def test_levenshtein_distance(self):
+        # copied from _testinternalcapi.test_edit_cost
+        # to also exercise the Python implementation
+
+        def CHECK(a, b, expected):
+            actual = traceback._levenshtein_distance(a, b, 4044)
+            self.assertEqual(actual, expected)
+
+        CHECK("", "", 0)
+        CHECK("", "a", 2)
+        CHECK("a", "A", 1)
+        CHECK("Apple", "Aple", 2)
+        CHECK("Banana", "B@n@n@", 6)
+        CHECK("Cherry", "Cherry!", 2)
+        CHECK("---0---", "------", 2)
+        CHECK("abc", "y", 6)
+        CHECK("aa", "bb", 4)
+        CHECK("aaaaa", "AAAAA", 5)
+        CHECK("wxyz", "wXyZ", 2)
+        CHECK("wxyz", "wXyZ123", 8)
+        CHECK("Python", "Java", 12)
+        CHECK("Java", "C#", 8)
+        CHECK("AbstractFoobarManager", "abstract_foobar_manager", 3+2*2)
+        CHECK("CPython", "PyPy", 10)
+        CHECK("CPython", "pypy", 11)
+        CHECK("AttributeError", "AttributeErrop", 2)
+        CHECK("AttributeError", "AttributeErrorTests", 10)
+        CHECK("ABA", "AAB", 4)
+
+    def test_levenshtein_distance_short_circuit(self):
+        from traceback import _substitution_cost, _MOVE_COST
+
+        if LEVENSHTEIN_DATA_FILE.is_file():
+            with LEVENSHTEIN_DATA_FILE.open("r") as f:
+                examples = json.load(f)
+        else:
+            print("recreating the missing data file")
+
+            from functools import cache
+            from random import choices, randrange
+
+            @cache
+            def levenshtein(a, b):
+                if not a or not b:
+                    return (len(a) + len(b)) * _MOVE_COST
+                option1 = levenshtein(a[:-1], b[:-1]) + _substitution_cost(a[-1], b[-1])
+                option2 = levenshtein(a[:-1], b) + _MOVE_COST
+                option3 = levenshtein(a, b[:-1]) + _MOVE_COST
+                return min(option1, option2, option3)
+
+            examples = set()
+            # Create a lot of non-empty examples, which should end up with a Gauss-like
+            # distribution for even costs (moves) and odd costs (case substitutions).
+            while len(examples) < 9990:
+                a = ''.join(choices("abcABC", k=randrange(1, 10)))
+                b = ''.join(choices("abcABC", k=randrange(1, 10)))
+                expected = levenshtein(a, b)
+                examples.add((a, b, expected))
+            # Create one empty case each for strings between 0 and 9 in length.
+            for i in range(10):
+                b = ''.join(choices("abcABC", k=i))
+                expected = levenshtein("", b)
+                examples.add(("", b, expected))
+            with LEVENSHTEIN_DATA_FILE.open("w") as f:
+                json.dump(sorted(examples), f, indent=2)
```

Comment:
This is a bit weird, running a test should not write into the test source directory. We should either move this generation into some helper script in Tools or as a standalone script, but not buried here in a branch in a test.

Reply:
Moved generation to Tools/scripts/.